app/jobs/create_virtual_objects_job.rb

```diff
@@ -0,0 +1,97 @@
+  # @param [Hash] params additional parameters that an Argo job may need
+  # @option params [String] :create_virtual_objects CSV string
+  def perform(bulk_action_id, params)
+    super
+
+    virtual_objects = VirtualObjectsCsvConverter.convert(csv_string: params[:create_virtual_objects])
+
+    # rubocop:disable Metrics/BlockLength
+    with_bulk_action_log do |log|
+      log.puts("#{Time.current} Starting #{self.class} for BulkAction #{bulk_action_id}")
+
+      parent_druids = parent_ids_from(virtual_objects)
+
+      # NOTE: We use this instead of `update_druid_count` because virtual object
+      #       creation does not use the `pids` form field.
+      bulk_action.update(druid_count_total: parent_druids.length)
+
+      find_problematic_druids!(parent_druids)
+
+      if problematic_druids.values.flatten.any?
+        log.puts("#{Time.current} #{UNAUTHORIZED_MESSAGE}: #{problematic_druids[:unauthorized].to_sentence}") if problematic_druids[:unauthorized].any?
+        log.puts("#{Time.current} #{NOT_FOUND_MESSAGE}: #{problematic_druids[:not_found].to_sentence}") if problematic_druids[:not_found].any?
+        bulk_action.update(druid_count_fail: problematic_druids.values.flatten.length)
+        # Remove problematic parent druids from list of objects to be operated upon
+        virtual_objects[:virtual_objects].reject! do |virtual_object|
+          problematic_druids.values.flatten.include?(virtual_object[:parent_id])
+        end
+        return if virtual_objects.values.flatten.empty?
+      end
+
+      begin
+        Dor::Services::Client.virtual_objects.create(virtual_objects)
+        bulk_action.update(druid_count_success: virtual_objects.length)
+        log.puts("#{Time.current} Successfully created virtual objects: #{parent_ids_from(virtual_objects).to_sentence}")
+      rescue Dor::Services::Client::UnexpectedResponse => e
+        errors = errors_from(e.message)
+
+        # Some exceptions are more unexpected than others
+        unless errors.respond_to?(:map)
+          Honeybadger.notify("Unexpected Dor::Services::Client error creating virtual objects: #{e.message}")
+          log.puts("#{Time.current} Terminating. An unexpected error occurred creating the batch of virtual objects: #{e.message}. \
+                   An alert has been sent. Counts for this batch will be off.")
+          return
+        end
+
+        errors.map do |error|
+          "Problem children for #{error.keys.first}: #{error.values.flatten.to_sentence}"
+        end
+        bulk_action.update(druid_count_fail: errors.length)
+        log.puts("#{Time.current} Creating some or all virtual objects failed because some objects are not combinable: #{errors.join('; ')}")
+      end
+
+      log.puts("#{Time.current} Finished #{self.class} for BulkAction #{bulk_action_id}")
+    end
+    # rubocop:enable Metrics/BlockLength
+  end
+
+  private
+
+  attr_reader :problematic_druids
+
+  def find_problematic_druids!(druids)
```

Comment:
I don't understand why we need to memoize and have an accessor as I only see this being invoked the one time.  How about just returning the data structure?